server/src/weaverbird/pipeline/formula_ast/eval.py

```diff
@@ -0,0 +1,155 @@
+import ast
+import tokenize
+from io import BytesIO
+from typing import Generator, Iterator
+
+from . import types
+
+
+class ASTError(Exception):
+    """Base exception for all AST errors"""
+
+
+class UnclosedColumnName(ASTError):
+    """Column name was not closed"""
+
+
+class EmptyColumnName(ASTError):
+    """Column name was not closed"""
+
+
+class UnexpectedToken(ASTError):
+    """Unexpected token"""
+
+
+class InvalidFormula(ASTError):
+    """Invalid formula"""
+
+
+class UnsupportedOperator(ASTError):
+    """Unsupported operator"""
+
+
+class UnsupportedExpression(ASTError):
+    """Unsupported expression"""
+
+
+class UnsupportedConstant(ASTError):
+    """Unsupported constant"""
+
+
+class FormulaParser:
+    def __init__(self, formula: str) -> None:
+        self._formula = formula
+        self._columns: dict[str, str] = {}
+
+    @staticmethod
+    def sanitize_string(s: str) -> str:
+        # Removing outer quotes
+        unquoted = ast.literal_eval(s)
+        # Escaping single quotes in the string
+        with_escaped_single_quotes = unquoted.replace('\'', r'\'')
+        # Wrapping the entire thing in single quotes
+        return f"'{with_escaped_single_quotes}'"
+
+    def _iterate_tokens(self, tokens: Iterator[tokenize.TokenInfo]) -> Generator[str, None, None]:
+        def next_token():
+            try:
+                return next(tokens)
+            except StopIteration:
+                return None
+
+        def parse_col_name(prev_token: tokenize.TokenInfo) -> str:
+            chunks = []
+            while tok := next_token():
+                if tok.type == tokenize.NAME:
+                    chunks.append(tok.string)
+                elif tok.type == tokenize.OP and tok.string == ']':
+                    if len(chunks) < 1:
+                        raise EmptyColumnName(f'Got an empty column name at {prev_token.start[1]}')
+                    col_name = f'__VQB_COL__{len(self._columns)}'
+                    self._columns[col_name] = ' '.join(chunks)
+                    return col_name
+                else:
+                    raise UnexpectedToken(f"Unexpected token {tok} in column name")
+                prev_token = tok
+            raise UnclosedColumnName(
+                f'Expected column to be closed near {prev_token.string} at {prev_token.start[1]}'
+            )
+
+        while token := next_token():
+            if token.type in (tokenize.ENCODING, tokenize.NEWLINE, tokenize.ENDMARKER):
+                continue
+            elif token.type == tokenize.STRING:
+                yield self.sanitize_string(token.string)
+            elif token.type == tokenize.OP and token.string == '[':
+                yield parse_col_name(token)
+            else:
+                yield token.string
```

Comment:
That seems a clever way for parsing `[some stuff]` as a unique `Name` called `some stuff`, yet it's a bit difficult to understand.
The prior implementation seemed more simple to me (I'm probably surely on this). It was about replacing `[...]` with a regex and the replacing back in `Name` elements (by walking the tree).

Do you have some arguments for one or the other approach?

Reply:
I feel more confident in this approach, because we will deny everything that is not deemed "an identifier" by python, so it'll avoid bad suprises later on when parsing and evaluating (for example there were some tests with `a column ^ name`, which could be interpreted as a valid formula with `^` being a power operator by some backends). 
And I'm really not a big fan of regexes :sweat_smile: 

Although now that I'm thinking about it, this approach does not allow for column names containing several consecutive whitespaces or starting with whitespaces, as they'll be stripped. when rebuilding the column identifier. So the best approach would probably be to just take everything between the start of the first and the end of the last `NAME` tokens. 

I'd rather keep this approach (and I'll definitely add comments everywhere), but if the majority prefers regexes, I'll change the implem :smile:

Reply:
The point about invalid characters and spaces are definitely valid :+1:  
Eventually, it shows that the decision is a product-led one: which column names should be supported, and which shouldn't.
Let's ask Tarek about this, and maybe the care team / product analytics, and decide afterwards! (In the meantime, keep the current approach)